### ISSUE TITLE
Get a list of symmetry operations for a space group

### ISSUE THREAD

**Issue by hakonanes** [pyxem/orix#87]
I can get all symmetry operations for space group 229 with diffpy.structure:

```python
>>> from diffpy.structure.spacegroups import GetSpaceGroup
>>> sg229 = GetSpaceGroup(229)
>>> list_sym_ops = sg229.symop_list[:sg229.num_primitive_sym_equiv]
>>> sym_ops = np.stack([op.R for op in list_sym_ops])
>>> print(sym_ops.shape)
(48, 3, 3)
>>> print(sym_ops[0])
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

Is it possible to get the same list from `orix.quaternion.symmetry.Symmetry`?

**Comment by hakonanes** [comment]
My question is more that is there some canonical way of getting the symmetry operations for a space group from the point group operations?

**Comment by dnjohnstone** [comment]
I *think* we only ever worried about point groups thus far because we weren't using the actual crystal structure anywhere before we added the diffpy dependency.

**Comment by hakonanes** [comment]
Yes, so actually what you should do when you define a phase is that you should state the space group instead of the point group?

The `diffpy.structure.Structure` object doesn't actually store any symmetry information at all, and I don't understand why. Should we add it to the `Phase` object ourselves? If so, we must update this whenever we change the point group or space group...

**Comment by hakonanes** [comment]
The point group is stored along with the spacegroup in diffpy.structure, so we just have to get a map from their notation to ours when we change the space group to update the point group.

The choice of space group is of course ambiguous when going the other way... I think whether a phase has a space group should depend on whether you explicitly set it or not:
* If you only set the point group, the phase does not have a space group.
* If a phase has a space group and you change the phase group, the space group is lost and you have to reset it.

**Comment by dnjohnstone** [comment]
So when we talk about a crystal structure it's definitely natural that the symmetry related with it is the space group. But when we talk about symmetry equivalence of crystal orientations it's only the proper point group that matters - I think that not everyone who might potentially use orix is as familiar as they should be with this so my ideal preference would be to give as much information as possible i.e. a phase should be explicitly related to: space group, point group, and proper point group

Does that make sense?

I think the thing with diffpy.structure.Structure not storing the space group information is that it was really developed to support pair distribution function analysis and in this case you probably want to refine the structure without symmetry by default and optionally impose symmetry operations as constraints if appropriate and/or later.

I think the ideal solution for us would be that there could be:

diffpy.structure.Structure

and the special case,

diffpy.structure.Crystal

Again though, we're flirting with changes that lead to making our own library for describing atomic structures and/or contributing to diffpy.structure.

We used to get all of this in diffsims by depending on pymatgen, which we stopped doing for various reasons, mostly that it just contains so much that we don't want.

**Comment by hakonanes** [comment]
I think the main question is how the `orix.quaternion.symmetry.Symmetry` class and its symmetry operations and `diffpy.structure.spacegroupmod.SpaceGroup` and its symmetry operations can be used in tandem without issues.

We could add the `SpaceGroup` object to the `Phase` object as a temporary solution and see...

**Comment by dnjohnstone** [comment]
Yeah that sounds good to me.

**Comment by hakonanes** [comment]
Okay, I'll set up a PR unless @pc494 has any objections or other comments.

**Comment by pc494** [comment]
Jumping in (albeit a little late), I wrote a space group to point group map for `diffsims`(which we are about to port to `orix`) that might do some of what you want @hakonanes? It's [here](https://github.com/pyxem/diffsims/blob/master/diffsims/utils/fundamental_zone_utils.py).

It slightly depends what we want to **do** with the spacegroup, but in general I'm moving towards asking the `diffpy` devs if we can get involved in some (small) ways. It may be that we start doing that and find out some of these things already exist.

**Comment by dnjohnstone** [comment]
So is the conclusion of this, given the positive experience of putting in the SpaceGroup representation to diffpy that we will now be thinking about porting stuff, particularly relating to crystal symmetry and crystallographic coordinate systems, to diffpy.structure?

**Comment by hakonanes** [comment]
> So is the conclusion of this, given the positive experience of putting in the SpaceGroup representation to diffpy, that we will now be thinking about porting stuff, particularly relating to crystal symmetry and crystallographic coordinate systems, to diffpy.structure?
> 
> If so - perhaps we should make a new issue and start to identify and make a list of suitable functionality?

i think so. although i havent used the symmetry stuff from orix directly yet, only the diffpy.structure SymOp point symmetry g-matrices, so i cannot now tell what the best action is. we should be aware of the fact that diffpy.structure supports python 2.7 as well as 3.7.

**Comment by dnjohnstone** [comment]
I suppose one thing we need to figure out is how/when we're going to push diffpy.structure for a new release...

**Comment by hakonanes** [comment]
Good point @dnjohnstone. It might be that we're heard if we say we just want a 3.0.2 patch release because of the space group representation?

Before suggesting a new minor release 3.1, I would like to at least get a comment from the developers on an issue I raised https://github.com/diffpy/diffpy.structure/issues/46 proposing to add direct and reciprocal structure matrices to the `Lattice` object upon initialization. (These are much used by EMsoft in their routines.)

**Comment by pc494** [comment]
> My question is more that is there some canonical way of getting the symmetry operations for a space group from the point group operations with orix?

Winding back, if you have the point group then

```
>>> from orix.quaternion.symmetry import Oh  
>>> Oh
```

Produces the 48 quaternions that you need. So then it's a matter of completing a  `to_matrix` method (see #79) and writing a `space_group --> point_group map`. Probably the easiest way to do this is to take the `space_group ---> proper_point_group` map in #90 and fix it up (plus maybe move it somewhere else in the code), as I believe (worth checking) that the improper components won't effect fundamental zone selection (so `Oh` and `O` will produce the same FZ).

**Comment by hakonanes** [comment]
Oh... In that case I suggest we get the `to_matrix()` etc. conversions in before 0.4, and actually wait a bit adding a `Phase.space_group` attribute.

**Comment by pc494** [comment]
Yep, I have some `diffsims` code that did some of this (as I'm sure I have mentioned before), I'll have a look through and say what's done/in need of doing in #79

**Comment by hakonanes** [comment]
Are there supposed to be 48 unique symmetry operations (`Rotation`) in the `Oh` `Symmetry` object? I find that only 24 of them are unique. Diffpy's `spacegroup` objects store 48 unique symmetry operations, defined by rotations AND translations. I guess the difference lies here?

```python
>>> from diffpy.structure.spacegroups import GetSpaceGroup
>>> import numpy as np
>>> from orix.symmetry import Oh
>>> sym_ops_orix = Oh.to_matrix()
>>> sym_ops_orix.shape
(48, 3, 3)
>>> np.unique(sym_ops_orix, axis=0).shape
(24, 3, 3)
>>> sg225 = GetSpaceGroup(225)
>>> sym_ops_dp = np.stack([i.R for i in sg225.symop_list])
>>> sym_ops_dp.shape
(192, 3, 3)
>>> np.unique(sym_ops_dp, axis=0).shape
(48, 3, 3)
```

Should we consider translations as well? If not I think I would like a `Phase.space_group` attribute from diffpy.structure.

**Comment by hakonanes** [comment]
Another argument *for* getting a `Phase.spacegroup` property is that, to get the asymmetric atomic units from `diffpy.structure.symmetryutilities.SymmetryConstraints().corepos`, [`SymmetryConstraints()`](https://github.com/diffpy/diffpy.structure/blob/master/src/diffpy/structure/symmetryutilities.py#L655) requires `diffpy.structure.Structure` and `diffpy.structure.spacegroupmod.SpaceGroup` objects upon initialization. Related routines, like expanding an asymmetric atomic unit with `symmetryutilities.expandPosition` to get atoms in the orbit, also requires both. I think it would be natural for the `Phase` object to store both.

**Comment by pc494** [comment]
I think I've been sold on this one; both this issue and #69 should probably be dealt with by adding `spacegroup` as an attribute of `Phase`. We could then also consider doing some other cute things like extracting point groups automatically. 

@hakonanes would you be happy to make an issue describing this enhancement to `Phase`? Happy to have it as `0.4.0` if you've got the time and drive but also `future` is just as good

**Comment by hakonanes** [comment]
> @hakonanes would you be happy to make an issue describing this enhancement to `Phase`? Happy to have it as `0.4.0` if you've got the time and drive but also `future` is just as good

Yes, 0.4 sounds good. I'll start working on it right away, as it is *really* convenient to use the `Phase` object with the orix `Symmetry` and `Vector3d` classes and the diffpy `Structure`, `Lattice` and `SpaceGroup` classes when building a reflector list for kikuchipy (I know there is some stuff in diffsims, might look into that at a later stage) like [EMsoft does it](https://github.com/EMsoft-org/EMsoft/blob/develop/Source/SEM/EMEBSDreflectors.f90).

**Comment by hakonanes** [comment]
I'll submit a WIP PR on Sunday with some API/organization questions.

**Comment by pc494** [comment]
This was closed by #104 (the Closes keyword doesn't understand "and" so doing this one manually)
